components/concept/push/src/main/java/mozilla/components/concept/push/PushProcessor.kt

```diff
@@ -6,68 +6,90 @@
 
 package mozilla.components.concept.push
 
-import android.support.annotation.CallSuper
+import android.support.annotation.VisibleForTesting
 
 /**
  * A push notification processor that handles registration and new messages from the [PushService] provided.
  * Starting Push in the Application's onCreate is recommended.
  */
-abstract class PushProcessor {
-
-    /**
-     * Initialize the PushProcessor before it
-     */
-    @CallSuper
-    open fun initialize() {
-        instance = this
-    }
+interface PushProcessor {
 
     /**
      * Start the push processor and starts any service associated.
      */
-    abstract fun start()
+    fun start()
 
     /**
      * Stop the push service and stops any service associated.
      */
-    abstract fun stop()
+    fun stop()
 
     /**
      * A new registration token has been received.
      */
-    abstract fun onNewToken(newToken: String)
+    fun onNewToken(newToken: String)
 
     /**
      * A new push message has been received.
      */
-    abstract fun onMessageReceived(message: PushMessage)
+    fun onMessageReceived(message: EncryptedPushMessage)
 
     /**
      * An error has occurred.
      */
-    abstract fun onError(error: Error)
-
-    companion object {
-        @Volatile
-        var instance: PushProcessor? = null
-        val requireInstance: PushProcessor
-            get() = instance ?: throw IllegalStateException(
-                "You need to call start() on your Push instance from Application.onCreate()."
-            )
-    }
+    fun onError(error: Error)
 }
 
 /**
- * A push message holds the information needed to pass the message on to the appropriate receiver.
+ * Provider class for the PushProcessor singleton.
  */
-data class PushMessage(val data: String)
+object PushProvider {
+
+    /**
+     * Initialize the PushProcessor that needs to be called before
+     */
+    fun install(processor: PushProcessor) {
+        instance = processor
+    }
+
+    @Volatile
+    private var instance: PushProcessor? = null
+
+    @VisibleForTesting
+    internal fun reset() {
+        instance = null
+    }
+    val requireInstance: PushProcessor
+        get() = instance ?: throw IllegalStateException(
+            "You need to call start() on your Push instance from Application.onCreate()."
```

Comment:
> I think a better home for the singleton is in the implementation. It's an implementation detail, after all. Conceivably you could have multiple types of push processors running at once, right? Against fcm, against a home-baked push service, whatever. Likely not in practice, but at the concept level you currently both allow it (by virtue of there being an abstract definition of a push processor), and discourage it (by defining a singleton).

Currently, multiple `PushService`'s can run in place but those services we likely don't have control over instantiating (in FCM's case it's init'd but a google plugin the app needs to add). Because of this, we have no way of passing an instance of a `PushProcessor` to the service first.

Using the `CrashReporter` component as an example, having a singleton provider seemed like the only viable solution to ensuring that we get all messages before the service starts.

<img width="797" alt="Screen Shot 2019-04-05 at 5 23 18 PM" src="https://user-images.githubusercontent.com/1370580/55725966-bff9b100-59dc-11e9-8d4c-7bee083c8ee5.png">

This is a communication chart of how each class talks to each other. (It assumes the app developer decided to disable FCM auto-start by default and starts it later by choice).

In the diagram, the `Push.kt` is an implementation of a `PushProcessor` that lives in the feature which receives all messages and tokens from a service, forwards it to the Rust `PushManager` to be decrypted. The message is then forwarded to the relevant listeners via the `MessageRouter`.